### ISSUE TITLE
Make presto JDBC driver work with Java 6

### ISSUE THREAD

**Issue by losipiuk** [prestodb/presto#3079]
Currently Presto JDBC requires Java 8 on the client side.
It is common that client applications use older version of Java (6 is still quite popular).
It would be nice to have version of JDBC driver which work with Java 6.

One option is to create another project - presto-jdbc-java6 - totally separate from Presto codebase. 
Such project would not share any code with driver in Java 8 driver in presto codebase.
The drawback of such approach is that we need to duplicate work as new JDBC related features are added to Presto. Implement them both in Java-6 and Java-8 codebase. Or even three times if we also want the Java-7 version.

Maybe a better option would be to generally move JDBC implementation out of presto codebase to separate repository presto-jdbc. There we could have separate maven projects for different java versions (6,7,8) - with shared modules containing common code.

On automation we would compile each of the projects using JDK in matching version.

The benefit of such approach - compared to leaving java8 implementation in presto repository - is that we have less code duplication. And as the JDBC project would not be very big - it should be maintainable even if we support multiple JDK versions.

Our (Teradata) intention is to do major enhancements to Presto JDBC driver. And we feel that it would be beneficial to the project if we targeted multiple JDK/JDBC versions.

What do you think?

**Comment by martint** [comment]
One challenge for moving the java8 implementation out of the presto repository is that it'd be impossible to do it without breaking up the repo into multiple projects. The verifier and benchmark driver have a dependency on the driver, which would result in circular dependencies between the two projects.

**Comment by ilfrin** [comment]
I see the problem you mention. I don't see any easy way around it. 
But that doesn't make the motivation for the change to be less important ;) Lets hope there's a way to get there.

In the long run I think it would be desirable to be able to allow different subprojects to allow connect to Presto. Breaking this up into separated modules with defined and separated responsibilities would be a way to implement that as you mention. 

For now, to unblock our work I think we should copy-paste the necessary code to a separate project and manually keep track of the changes made to this code on both sides (yuck!). 

In the background I propose we start fixing the problem more systematically. 
**The problem** is we have a set of model classes on the wire and some logic to manipulate them on each end. 

**One option** is to copy-paste it to every project that needs to talk to Presto, I would strongly opt we throw this out the window. 
**Second option** would be to break the Presto project into subprojects, have a presto-cli-base that is java 6 compatible and other projects (including Presto, verifier, jdbc, ...) depend on it, that should allow to overcome the circular dependencies. This requires a lot of effort and rearranges things a lot. Maybe too much and not worth the effort. Maybe on the other hand there are other reasons to do it, eg. some of the Presto sub-modules today should be external tools depending on Presto, not inside Presto, I don't know, haven't analyzed it all, just speculating that there may be more motivation for such a separation. The net result could be a cleanly separated set of projects. Then again it may complicate the daily development too much if there is a lot of work spanning all potential sub-projects. 
**Third option** we could consider is to describe the protocol in a way that allows the clients the automatically generate the necessary model. Depending on the implementation choice the Presto listener side keeps it's hand-coded version or wraps the generated classes with some more custom validation logic, which is today present there. Leveraging current implementation choice (JAX RS) we could express the protocol via a WADL file, problem we would have is that Presto today send meta information in headers and handle them by hand, WADL supports header information but we'd need to change the way we handle it. Another approach would be to reimplement the meta-information to embed it in the JSON. Last (not least) way to do would be to start using something like GRPC, or sth else instead of JAX RS. 

Do you have any other ideas? Which of those would you prefer?

**Comment by electrum** [comment]
The client protocol is rather simple and stable. It has to be, since there are implementations in many languages. We don't worry about duplicate model definitions in other languages, so I don't think it is a big deal for the Java 6 JDBC driver.

One advantage of a separate driver is that it can be smaller by using the JDK HTTP client, etc., rather than all the dependencies that the one in Presto uses. (Not that anyone has ever cared about the driver size, and possibly never will, but mentioning it anyway)

**Comment by ilfrin** [comment]
The client data structures is pretty stable at this point. Maybe that's the reason for it being stable that there was not too much of development that mandated changes to the interface. Hopefully there will be few changes to it during the implementation of new jdbc features and we'll live with it. Like I said we will start with a separate project copy-pasting things. 

Not setting things in stone, I wanted to test the ground and just see how you feel about moving that situation forward, which of the options and their variations you view as optimal. If this is to happen somewhere in the future I'd like our thinking to start aligning as early in the process as possible. 

Thanks.

**Comment by losipiuk** [comment]
Closing this one. There is no work planned in near future. And there is closed-source alternative which works under Java6
